[AMDGPU] Reschedule loads in clauses to improve throughput

perlfu opened the pull request:
After clauses are formed their internal loads can be reordered to facilitate some additional opportunities for overlapping computation.
This late stage rescheduling causes no change register pressure.

perlfu commented:
Implementation is not entirely complete; however, I would be interested in feedback, specifically if we think this is "OK" from a memory model perspective.
My opinion is currently "yes", with a few additional checks on address spaces, cache bypass flags and atomicity that are currently missing from the code.
This draft only targets MIMG, but I think it makes sense for all VMEM clauses.

jayfoad commented:
Can you explain what kind of rescheduling you are doing? Why would the regular scheduler not already have put the loads into an optimal order?

jayfoad commented:
Ah, `ReorderWhileClustering` is a `MachineScheduler` option. I had forgotten about that (even though I reviewed it).

> the implicit order drives scheduling when other factors are discounted -- e.g. when a sequence of loads depend on a value computed by a VALU immediately beforehand they are all in the stall state, so the implicit order of the cluster tends to determine schedule order.

Yeah, I think that is part of the scheduler's "do no harm" policy, i.e. don't change anything unless you have a good reason to do so.

For "This patch (reorder after clustering+RA)" what am I supposed to be looking at? Why is it better than the other examples?

Stepping back a bit, implementing clustering by adding edges to the DAG seems like a bad idea, since DAG edges are directional so that forces you to pick an ordering up-front. It would be better if clustering only caused the loads to be scheduled adjacently, and all the other scheduler heuristics could still influence which order they go in.

perlfu commented:
I am not against enabling `ReorderWhileClustering`; however, it will potentially change RA -- so could have some unexpected negatives.
The point of my examples is that "reorder after clustering+RA" can take the "default output" and achieve results similar to or better than `ReorderWhileClustering` by reordering, with no impact to existing RA or schedules.
I say *better* because each successive waitcnt is only waiting for 1-2 more outstanding loads, rather than 7+ in the default case.
This might seem not useful at first because only a single VALU is issued before potentially another wait, but if we imagine (n) waves in flight, then that means (n) VALUs that can issued.

I have not tested extensively, but I have the impression that this kind of reordering can, in some cases, give a few percentage points performance improvement.

perlfu committed:
[AMDGPU] Reschedule loads in clauses to improve throughput

After clauses are formed their internal loads can be reordered
to facilitate some additional opportunities for overlapping
computation.
This late stage rescheduling causes no change in register pressure.

arsenm commented on llvm/lib/Target/AMDGPU/SIPostRABundler.cpp:
```diff
@@ -129,6 +134,141 @@ bool SIPostRABundler::canBundle(const MachineInstr &MI,
+    if (!II->getNumExplicitDefs() || II->isKill() ||
+        llvm::any_of(II->memoperands(), [&](const MachineMemOperand *MMO) {
+          return MMO->isAtomic() || MMO->isVolatile();
+        })) {
```
hasOrderedMemRef?

perlfu commented:
- Rebase
- Address reviewer comments

arsenm commented on llvm/lib/Target/AMDGPU/SIPostRABundler.cpp:
Avoid double map lookup

perlfu replied:
Done

arsenm commented on llvm/lib/Target/AMDGPU/SIPostRABundler.cpp:
Need to skip debug instructions?

perlfu replied:
I have added better handling of meta/debug instructions

perlfu committed:
- Test updates

perlfu committed:
- Test updates

ruiling commented on llvm/lib/Target/AMDGPU/SIPostRABundler.cpp:
```diff
@@ -137,6 +142,143 @@ bool SIPostRABundler::canBundle(const MachineInstr &MI,
+        LLVM_DEBUG(dbgs() << "  Trivially OK\n");
+        break;
+      }
+
+      Register DefReg = getDef(*MI);
+      bool DefRegHasUse = false;
+      for (auto SearchIt = std::next(It);
+           SearchIt != Schedule.end() && !DefRegHasUse; ++SearchIt)
+        DefRegHasUse = SearchIt->first->readsRegister(DefReg, TRI);
+      if (DefRegHasUse) {
```
To reorder two instructions (inst2 appears after inst1 in the old order)
```
inst1
inst2
```
I think we need to detect three kinds of conflicts theoretically:
1. inst2 read inst1's def
2. inst2 write to Inst1's src
3. inst2 write to Inst1's def

I think your code handles the second conflict. For the first kind, I am not sure do we already make it impossible somewhere else? But for the third case, I think we still need to handle them. I am not sure whether doing the check through `regunits()` would simplify the logic.

perlfu replied:
Thanks for looking at this.

In `canBundle()`, the slightly misleadingly named `isDependentLoad()` ensures that the explicit operands of any instruction added to the bundle do not overlap with any of the `defs` already in the bundle.
I believe this prevents both (1) and (3), specifically because both of these would introduce a `s_wait` into the bundle.
My new code handles (2) because read followed by write would not block bundling, but does limit reordering.

The only potential issue I can see related to (3) is for implicit defs, as these are not currently considered.
i.e.
```
$reg0 = inst1, implicit-def $reg0_reg1
$reg1 = inst2
```
Reordering these causes a kind of semantic issue for liveness, but no practical issue for instruction correctness.

ruiling replied:
Make sense. It would be better to add some comment here like: Other register dependencies are handled by bundle checking earlier. This would save some time in case others have the same confusion.

ruiling reviewed (commented):
I think it is ok to have this. The change basically looks good.

perlfu committed:
- Test updates

ruiling commented on llvm/lib/Target/AMDGPU/SIPostRABundler.cpp:
```diff
@@ -137,6 +142,143 @@ bool SIPostRABundler::canBundle(const MachineInstr &MI,
 }
 
+static Register getDef(MachineInstr &MI) {
+  assert(MI.getNumExplicitDefs() > 0);
```
Shouldn't we assert `(MI.getNumExplicitDefs()` is exactly `1`? given that we only handle the first def.

perlfu replied:
Actually if this is `== 1` we get test failures.
This is an inherited limitation from the original code.
It does create a risk as we are relying on `Defs` modelling all defs.
So I have adjusted all the code to now handle multiple defs, and this function is removed entirely.

ruiling commented on llvm/lib/Target/AMDGPU/SIPostRABundler.cpp:
```diff
@@ -242,6 +386,7 @@ bool SIPostRABundler::run(MachineFunction &MF) {
           BundleUsedRegUnits.reset();
         }
 
+        reorderLoads(MBB, BundleStart, Next);
```
Instead of modifying BundleStart implicitly, would it be better to return the new start and assign to BundleStart after return from the call?

perlfu replied:
Done.

ruiling reviewed (approved):
LGTM, just some nits.

perlfu committed:
- Make pass handle multiple def instructions
- Address reviewer comments

ruiling reviewed (approved):
Thanks for fixing the multiple defs case. Still LGTM
